From aea882206c575195c367469d7e449d7cb5cbddd8 Mon Sep 17 00:00:00 2001
From: Tim Besard <tim.besard@gmail.com>
Date: Sun, 27 Sep 2026 14:55:20 +0200
Subject: [PATCH] UnreachablesToReturns: don't delete the entry block

When a kernel is found to always hit undefined behavior (e.g. because it
branches on undef), LLVM reduces its body to a single unreachable instruction.
The entry block then ended in an unreachable and had no predecessors, so it was
deleted like any other such block, turning the kernel into a declaration. The
work-group function then failed to inline it, and kept calling it after the
kernel was erased, crashing later passes. Convert the unreachable in the entry
block to a return instead.

(cherry picked from commit 9b740292e265eb3fc7f72d821ddcbc30a1ae1f68)
---
 lib/llvmopencl/UnreachablesToReturns.cc    |   7 +-
 tests/regression/CMakeLists.txt            |   3 +
 tests/regression/test_unreachable_kernel.c | 150 +++++++++++++++++++++
 3 files changed, 158 insertions(+), 2 deletions(-)
 create mode 100644 tests/regression/test_unreachable_kernel.c

diff --git a/lib/llvmopencl/UnreachablesToReturns.cc b/lib/llvmopencl/UnreachablesToReturns.cc
index f7420788c..222075957 100644
--- a/lib/llvmopencl/UnreachablesToReturns.cc
+++ b/lib/llvmopencl/UnreachablesToReturns.cc
@@ -101,8 +101,11 @@ static bool convertUnreachablesToReturns(Function &F) {
       LLVM_DEBUG(dbgs() << "UNREACHABLE found: replacing Inst in "
                         << F.getName().str() << "\n");
       // this can happen when inlining functions which have unreachable Inst
-      // we end up with a BB with 0 predecessors and a single unreachable
-      if (BB.hasNPredecessors(0))
+      // we end up with a BB with 0 predecessors and a single unreachable.
+      // The entry block has no predecessors either, but deleting it would
+      // turn the function into a declaration (e.g. when the whole kernel is
+      // known to be unreachable), so convert it to a return instead.
+      if (BB.hasNPredecessors(0) && &BB != &F.getEntryBlock())
         PendingDeletableBBs.push_back(&BB);
       else
         PendingUnreachableInst.push_back(UI);
diff --git a/tests/regression/CMakeLists.txt b/tests/regression/CMakeLists.txt
index 36137e9f6..5c50f524d 100644
--- a/tests/regression/CMakeLists.txt
+++ b/tests/regression/CMakeLists.txt
@@ -29,6 +29,7 @@ set(C_PROGRAMS_TO_BUILD test_assign_loop_variable_to_privvar_makes_it_local
      test_assign_loop_variable_to_privvar_makes_it_local_2
      test_llvm_segfault_issue_889
      test_issue_1711 test_issue_1794 test_issue_1747 test_issue_1958
+     test_unreachable_kernel
      test_unreachable_in_infinite_loop
      test_rematerialized_alloca_load_with_outside_pr_users
      test_peeled_wi_global_id
@@ -169,6 +170,8 @@ add_test_pocl(NAME "regression/test_issue_1747" COMMAND "test_issue_1747" "0" "$
 
 add_test_pocl(NAME "regression/test_issue_1958" WORKITEM_HANDLER "loops;loopvec;cbs;" COMMAND "test_issue_1958" LABELS ${SPIRV_LABELS})
 
+add_test_pocl(NAME "regression/test_unreachable_kernel" WORKITEM_HANDLER "loops;loopvec;cbs;" COMMAND "test_unreachable_kernel" LABELS ${SPIRV_LABELS})
+
 add_test_pocl(NAME "regression/unreachable_in_infinite_loop" WORKITEM_HANDLER "loops;loopvec;cbs;" COMMAND "test_unreachable_in_infinite_loop")
 
 add_test_pocl(NAME "regression/peeled_wi_global_id" WORKITEM_HANDLER "loopvec;cbs;" COMMAND "test_peeled_wi_global_id" LABELS ${SPIRV_LABELS})
diff --git a/tests/regression/test_unreachable_kernel.c b/tests/regression/test_unreachable_kernel.c
new file mode 100644
index 000000000..dd8efffe4
--- /dev/null
+++ b/tests/regression/test_unreachable_kernel.c
@@ -0,0 +1,150 @@
+// Copyright (c) 2026 PoCL developers
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/*
+  A kernel that branches on an undefined condition (found while reducing a
+  Julia kernel with spirv-reduce, which introduced the undef;
+  https://github.com/JuliaGPU/OpenCL.jl/issues/509):
+
+      kernel void kernel() {
+        if (undef)
+          callee();
+      }
+
+  Branching on undef is undefined behavior, so LLVM reduces the kernel body to
+  a single unreachable instruction. ConvertUnreachablesToReturns then deleted
+  that entry block, as it deletes other predecessor-less blocks ending in an
+  unreachable, which turned the kernel into a declaration that the work-group
+  function could not inline, leaving a call to the deleted kernel behind.
+
+  The crash happened when building the work-group function, so completing the
+  build and launch is the test. Test source (test_unreachable_kernel.spv):
+
+               OpCapability Kernel
+               OpCapability Addresses
+               OpMemoryModel Physical64 OpenCL
+               OpEntryPoint Kernel %kernel "kernel"
+       %void = OpTypeVoid
+       %bool = OpTypeBool
+         %fn = OpTypeFunction %void
+      %undef = OpUndef %bool
+     %callee = OpFunction %void None %fn
+          %1 = OpLabel
+               OpReturn
+               OpFunctionEnd
+     %kernel = OpFunction %void None %fn
+      %entry = OpLabel
+               OpBranchConditional %undef %then %join
+       %then = OpLabel
+          %2 = OpFunctionCall %void %callee
+               OpBranch %join
+       %join = OpLabel
+               OpReturn
+               OpFunctionEnd
+*/
+
+#include "pocl_opencl.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#ifndef _WIN32
+#include <unistd.h>
+#endif
+
+#define CHECK_ERROR(err)                                                       \
+  if (err != CL_SUCCESS) {                                                     \
+    printf("OpenCL Error %d at %s:%d\n", err, __FILE__, __LINE__);             \
+    return 1;                                                                  \
+  }
+
+int main(int argc, char **argv) {
+  cl_uint platform_index = argc > 1 ? (cl_uint)atoi(argv[1]) : 0;
+  cl_int err;
+
+#ifndef _WIN32
+  // Fail quickly instead of tripping the much larger ctest timeout in case
+  // a regressed build hangs.
+  alarm(300);
+#endif
+
+  cl_uint num_platforms;
+  CHECK_ERROR(clGetPlatformIDs(0, NULL, &num_platforms));
+  if (platform_index >= num_platforms) {
+    printf("Platform index %u out of range\n", platform_index);
+    return 1;
+  }
+  cl_platform_id *platforms = malloc(sizeof(cl_platform_id) * num_platforms);
+  CHECK_ERROR(clGetPlatformIDs(num_platforms, platforms, NULL));
+  cl_platform_id platform = platforms[platform_index];
+  free(platforms);
+
+  cl_device_id device;
+  CHECK_ERROR(clGetDeviceIDs(platform, CL_DEVICE_TYPE_ALL, 1, &device, NULL));
+
+  if (!poclu_device_supports_il(device, "SPIR-V_1.0")) {
+    printf("SKIP: The test requires support for SPIR-V 1.0\n");
+    return 77;
+  }
+
+  cl_context context = clCreateContext(NULL, 1, &device, NULL, NULL, &err);
+  CHECK_ERROR(err);
+  cl_queue_properties props[] = {0};
+  cl_command_queue queue =
+      clCreateCommandQueueWithProperties(context, device, props, &err);
+  CHECK_ERROR(err);
+
+  FILE *f = fopen(SRCDIR "/test_unreachable_kernel.spv", "rb");
+  if (!f) {
+    printf("Failed to open test_unreachable_kernel.spv\n");
+    return 1;
+  }
+  fseek(f, 0, SEEK_END);
+  size_t size = ftell(f);
+  fseek(f, 0, SEEK_SET);
+  unsigned char *binary = malloc(size);
+  if (fread(binary, 1, size, f) != size) {
+    printf("Failed to read SPIR-V\n");
+    free(binary);
+    fclose(f);
+    return 1;
+  }
+  fclose(f);
+
+  cl_program program = clCreateProgramWithIL(context, binary, size, &err);
+  CHECK_ERROR(err);
+  CHECK_ERROR(clBuildProgram(program, 1, &device, NULL, NULL, NULL));
+
+  cl_kernel kernel = clCreateKernel(program, "kernel", &err);
+  CHECK_ERROR(err);
+
+  size_t global_size = 16, local_size = 8;
+  CHECK_ERROR(clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &global_size,
+                                     &local_size, 0, NULL, NULL));
+  CHECK_ERROR(clFinish(queue));
+
+  clReleaseKernel(kernel);
+  clReleaseProgram(program);
+  clReleaseCommandQueue(queue);
+  clReleaseContext(context);
+  free(binary);
+
+  printf("OK\n");
+  return 0;
+}
